File: server/logGenerator.ts
/**
 * RouteLedger - ELD / RODS Daily Log Sheet Generator (TypeScript)
 * Maps day schedule events into FMCSA 49 CFR § 395.8 grid lines, remarks, and headers.
 */
import { DayPlan, ELDDailyLogSheet, ELDGraphSegment, CarrierInfo } from '../src/types.ts';

export function generateELDLogs(
  days: DayPlan[],
  carrierInfo?: Partial<CarrierInfo>
): ELDDailyLogSheet[] {
  const carrierName = carrierInfo?.carrier_name || 'Apex Freight Systems LLC';
  const mainOffice = carrierInfo?.main_office || '4200 Logistics Blvd, Richmond, VA 23230';
  const homeTerminal = carrierInfo?.home_terminal || 'Richmond Terminal, VA';
  const driverName = carrierInfo?.driver_name || 'John R. Miller';
  const vehicleNumber = carrierInfo?.vehicle_number || 'TRK-4089';
  const trailerNumber = carrierInfo?.trailer_number || 'TLR-8821';
  const shippingDoc = carrierInfo?.shipping_doc || 'BOL-77341 / Consumer Freight';
  const coDriver = carrierInfo?.co_driver || 'Solo Driver';

  return days.map((d) => {
    const graphSegments: ELDGraphSegment[] = [];

    for (const ev of d.events) {
      const dStart = new Date(ev.start_time);
      const dEnd = new Date(ev.end_time);

      const hStart = dStart.getUTCHours();
      const mStart = dStart.getUTCMinutes();
      const hEnd = dEnd.getUTCHours();
      const mEnd = dEnd.getUTCMinutes();

      const timeStartStr = `${String(hStart).padStart(2, '0')}:${String(mStart).padStart(2, '0')}`;
      let timeEndStr = `${String(hEnd).padStart(2, '0')}:${String(mEnd).padStart(2, '0')}`;

      let startFraction = (hStart + mStart / 60) / 24;
      let endFraction = (hEnd + mEnd / 60) / 24;

      if (endFraction === 0 && (hEnd !== hStart || mEnd !== mStart)) {
        endFraction = 1.0;
        timeEndStr = '24:00';
      }

      graphSegments.push({
        event_id: ev.id,
        duty_status: ev.duty_status,
        start_time: timeStartStr,
        end_time: timeEndStr,
        start_fraction: parseFloat(startFraction.toFixed(4)),
        end_fraction: parseFloat(endFraction.toFixed(4)),
        duration_hours: ev.duration_hours,
        location_name: ev.location_name,
        description: ev.description,
      });
    }

    return {
      day_number: d.day_number,
      date: d.date,
      carrier_name: carrierName,
      main_office_address: mainOffice,
      home_terminal_address: homeTerminal,
      driver_name: driverName,
      co_driver: coDriver,
      truck_tractor_number: vehicleNumber,
      trailer_number: trailerNumber,
      shipping_documents: shippingDoc,
      total_miles_driving_today: d.total_miles_driving_today,
      total_hours: d.total_hours,
      totals: {
        off_duty_hours: d.off_duty_hours,
        sleeper_berth_hours: d.sleeper_berth_hours,
        driving_hours: d.driving_hours,
        on_duty_not_driving_hours: d.on_duty_not_driving_hours,
        total_on_duty_hours: d.total_on_duty_hours,
      },
      graph_segments: graphSegments,
      remarks: d.remarks,
      certified: true,
      certification_statement: `I hereby certify that my data entries and my record of duty status for ${d.date} are true and correct. — ${driverName}`,
    };
  });
}

File: backend/trips/services/log_generator.py
"""
RouteLedger - ELD/RODS Daily Log Generator
Produces FMCSA 49 CFR § 395.8 compliant Driver's Daily Log data structures.
Splits multi-day journeys into individual 24-hour log sheets with exact status segments, totals, and remarks.
"""

from typing import List, Dict, Any, Optional


class ELDLogGenerator:
    @staticmethod
    def generate_daily_logs(
        days: List[Dict[str, Any]],
        carrier_info: Optional[Dict[str, Any]] = None,
    ) -> List[Dict[str, Any]]:
        """
        Takes day-by-day schedules from HOSPlanner and attaches carrier, vehicle,
        shipping, and driver certification metadata conforming to FMCSA RODS layout.
        """
        info = carrier_info or {}
        carrier_name = info.get("carrier_name", "Apex Freight Systems LLC")
        main_office = info.get("main_office", "4200 Logistics Blvd, Richmond, VA 23230")
        home_terminal = info.get("home_terminal", "Richmond Terminal, VA")
        driver_name = info.get("driver_name", "John R. Miller")
        vehicle_num = info.get("vehicle_number", "TRK-4089")
        trailer_num = info.get("trailer_number", "TLR-8821")
        shipping_doc = info.get("shipping_doc", "BOL-77341 / Consumer Goods")
        co_driver = info.get("co_driver", "N/A - Solo Driver")

        daily_logs: List[Dict[str, Any]] = []

        for d in days:
            events = d.get("events", [])
            remarks = d.get("remarks", [])

            # Generate SVG-ready time segments for the 24-hour graph
            # 00:00 is x=0%, 24:00 is x=100%
            graph_segments = []
            for ev in events:
                start_iso = ev["start_time"]
                end_iso = ev["end_time"]

                # Extract hour & minute
                # ISO: YYYY-MM-DDTHH:MM:SS...
                t_start = start_iso.split("T")[1][:5]
                t_end = end_iso.split("T")[1][:5]

                h_s, m_s = map(int, t_start.split(":"))
                h_e, m_e = map(int, t_end.split(":"))

                start_fraction = (h_s + m_s / 60.0) / 24.0
                end_fraction = (h_e + m_e / 60.0) / 24.0
                # Handle midnight wrap (24:00)
                if end_fraction == 0.0 and (h_e != h_s or m_e != m_s):
                    end_fraction = 1.0

                graph_segments.append({
                    "event_id": ev.get("id"),
                    "duty_status": ev["duty_status"],
                    "start_time": t_start,
                    "end_time": t_end,
                    "start_fraction": round(start_fraction, 4),
                    "end_fraction": round(end_fraction, 4),
                    "duration_hours": ev["duration_hours"],
                    "location_name": ev.get("location_name", ""),
                    "description": ev.get("description", ""),
                })

            daily_logs.append({
                "day_number": d["day_number"],
                "date": d["date"],
                "carrier_name": carrier_name,
                "main_office_address": main_office,
                "home_terminal_address": home_terminal,
                "driver_name": driver_name,
                "co_driver": co_driver,
                "truck_tractor_number": vehicle_num,
                "trailer_number": trailer_num,
                "shipping_documents": shipping_doc,
                "total_miles_driving_today": d.get("total_miles_driving_today", 0.0),
                "total_hours": d.get("total_hours", 24.0),
                "totals": {
                    "off_duty_hours": d.get("off_duty_hours", 0.0),
                    "sleeper_berth_hours": d.get("sleeper_berth_hours", 0.0),
                    "driving_hours": d.get("driving_hours", 0.0),
                    "on_duty_not_driving_hours": d.get("on_duty_not_driving_hours", 0.0),
                    "total_on_duty_hours": d.get("total_on_duty_hours", 0.0),
                },
                "graph_segments": graph_segments,
                "remarks": remarks,
                "certified": True,
                "certification_statement": (
                    f"I hereby certify that my data entries and my record of duty status "
                    f"for {d['date']} are true and correct. — {driver_name}"
                ),
            })

        return daily_logs
